File: java/src/main/java/neetcode/twopointers/p0042_trappingrainwater/Solution.java
package neetcode.twopointers.p0042_trappingrainwater;

public class Solution {
    /**
     * Returns the total amount of water that can be trapped after raining,
     * given an elevation map where each bar has width 1.
     *
     * Approach: converging two pointers maintaining the max seen on each side.
     * The key insight is that we only need to know one side's max precisely
     * to compute the water at the current cell — the other side is guaranteed
     * to have a taller (or equal) wall by the pointer comparison itself.
     *
     * Time:  O(n)
     * Space: O(1)
     */
    public int trap(int[] height) {
        int left = 0;
        int right = height.length - 1;
        int leftMax = 0;
        int rightMax = 0;
        int total = 0;

        while (left < right) {
            // Process the shorter side: its max is the binding constraint
            if (height[left] < height[right]) {
                if (height[left] >= leftMax) {
                    leftMax = height[left];        // new tallest wall on the left
                } else {
                    total += leftMax - height[left];   // water sits above this cell
                }
                left++;
            } else {
                if (height[right] >= rightMax) {
                    rightMax = height[right];
                } else {
                    total += rightMax - height[right];
                }
                right--;
            }
        }
        return total;
    }
}

File: python/src/neetcode/twopointers/p0042_trappingrainwater/solution.py
from typing import List


class Solution:
    def trap(self, height: List[int]) -> int:
        """
        Returns the total water trapped after raining.

        Approach: converging two pointers with running max on each side.
        Process the side with the shorter current bar — its max is the
        binding constraint. Update the running max or accumulate water.

        Time:  O(n)
        Space: O(1)
        """
        left, right = 0, len(height) - 1
        leftMax, rightMax = 0, 0
        total = 0

        while left < right:
            if height[left] < height[right]:
                if height[left] >= leftMax:
                    leftMax = height[left]
                else:
                    total += leftMax - height[left]
                left += 1
            else:
                if height[right] >= rightMax:
                    rightMax = height[right]
                else:
                    total += rightMax - height[right]
                right -= 1

        return total
